tests/test_data_utils/test_downsample_cells.py

```diff
@@ -1,98 +1,150 @@
             n_samples=n_samples, stratify=False
         )
         self.assertTrue(
-            all(df_downsampled.groupby(
-                self.annotation
+            all(df_downsampled_single.groupby(
+                self.annotation_single
             ).size().values <= n_samples)
         )
-
-    def test_downsample_with_stratify_without_rand(self):
-        """Downsample with stratify without random"""
-        n_samples = 20
-        df_downsampled = downsample_cells(
-            self.df, self.annotation,
-            n_samples=n_samples, stratify=True
+        # For multi obs, group by both columns without combining them
+        self.assertTrue(
+            all(df_downsampled_multi.groupby(
+                self.annotation_multi
+            ).size().values <= n_samples)
         )
-        self.assertEqual(df_downsampled.shape[0], n_samples)
 
-    def test_downsample_with_stratify_with_rand(self):
-        """Downsample with stratify with random"""
-        n_samples = 20
-        df_downsampled = downsample_cells(
-            self.df, self.annotation,
-            n_samples=n_samples, stratify=True, rand=True
-        )
-        self.assertEqual(df_downsampled.shape[0], n_samples)
+    def check_stratified_counts(
+            self, downsampled_df, columns, expected_counts):
+        """
+        Helper method to check stratified counts for both single and
+        multi annotations.
+        """
+        combined_col = '_'.join(columns)
 
-    def test_downsample_no_n_samples(self):
-        """Downsample without n_samples should return the original dataframe"""
-        df_downsampled = downsample_cells(self.df, self.annotation)
-        pd.testing.assert_frame_equal(self.df, df_downsampled)
+        # Ensure the combined column exists in the downsampled_df
+        if combined_col not in downsampled_df.columns:
+            downsampled_df[combined_col] = downsampled_df[columns].apply(
+                lambda row: '_'.join(row.values.astype(str)), axis=1
+            )
 
-    def test_stratification_frequency(self):
-        """Test the frequency of stratification"""
-        n_samples = 10
+        actual_counts = downsampled_df[combined_col].value_counts()
+
+        for key, expected_count in expected_counts.items():
+            self.assertEqual(expected_count, actual_counts.get(key, 0))
+
+        # Remove the combined column after checks
+        if combined_col in downsampled_df.columns and len(columns) > 1:
+            # Only drop if it's a multi annotation
+            downsampled_df.drop(columns=combined_col, inplace=True)
 
-        # Downsampling with stratifying
+    def test_stratification_frequency_single_obs(self):
+        """Test the frequency of stratification for single annotation."""
+        n_samples = 10
         df_downsampled_stratified = downsample_cells(
-            self.df, self.annotation,
+            self.df_single_fixed, self.annotation_single,
```

Comment:
@fangliu117, would you like to use df_single_random instead which has 100 cells? This way we can see that downsampling actually took place. you can keep this test as well as a corner case.

Reply:
yes, df_single_random was applied.